cli/cmd/plugin/plugin.go

```diff
@@ -0,0 +1,234 @@
+
+	// SystemCmdGroup are system commands.
+	SystemCmdGroup CmdGroup = "System"
+
+	// VersionCmdGroup are version commands.
+	VersionCmdGroup CmdGroup = "Version"
+
+	// AdminCmdGroup are admin commands.
+	AdminCmdGroup CmdGroup = "Admin"
+
+	// TestCmdGroup is the test command group.
+	TestCmdGroup CmdGroup = "Test"
+
+	// ExtraCmdGroup is the extra command group.
+	ExtraCmdGroup CmdGroup = "Extra"
+)
+
+// PluginDescriptor describes a plugin binary.
+type PluginDescriptor struct {
+	// Name is the name of the plugin.
+	Name string `json:"name" yaml:"name"`
+
+	// Description is the plugin's description.
+	Description string `json:"description" yaml:"description"`
+
+	// Version of the plugin. Must be a valid semantic version https://semver.org/
+	Version string `json:"version" yaml:"version"`
+
+	// BuildSHA is the git commit hash the plugin was built with.
+	// TODO(stmcginnis): Update Makefile to set build info with LDFLAG.
+	BuildSHA string `json:"buildSHA" yaml:"buildSHA"`
+
+	// Digest is the SHA256 hash of the plugin binary.
+	Digest string `json:"digest" yaml:"digest"`
+
+	// Command group for the plugin.
+	Group CmdGroup `json:"group" yaml:"group"`
+
+	// DocURL for the plugin.
+	DocURL string `json:"docURL,omitempty" yaml:"docURL,omitempty"`
+
+	// Hidden tells whether the plugin should be hidden from the help command.
+	Hidden bool `json:"hidden,omitempty" yaml:"hidden,omitempty"`
+
+	// CompletionType determines how command line completion will be determined.
+	CompletionType PluginCompletionType `json:"completionType" yaml:"completionType"`
+
+	// CompletionArgs contains the valid command line completion values if `CompletionType`
+	// is set to `StaticPluginCompletion`.
+	CompletionArgs []string `json:"completionArgs,omitempty" yaml:"completionArgs,omitempty"`
+
+	// CompletionCommand is the command to call from the plugin to retrieve a list of
+	// valid completion nouns when `CompletionType` is set to `DynamicPluginCompletion`.
+	CompletionCommand string `json:"completionCmd,omitempty" yaml:"completionCmd,omitempty"`
+
+	// Aliases are other text strings used to call this command
+	Aliases []string `json:"aliases,omitempty" yaml:"aliases,omitempty"`
+
+	// InstallationPath is a relative installation path for a plugin binary.
+	// E.g., cluster/v0.3.2@sha256:...
+	InstallationPath string `json:"installationPath"`
+
+	// Discovery is the name of the discovery from where
+	// this plugin is discovered.
+	Discovery string `json:"discovery"`
+
+	// Scope is the scope of the plugin. Stand-Alone or Context
+	Scope string `json:"scope"`
+}
+
+// NewPlugin creates an instance of Plugin.
+func NewPlugin(descriptor *PluginDescriptor) (*Plugin, error) {
+	p := &Plugin{
+		Cmd: &cobra.Command{
+			Use:     descriptor.Name,
+			Short:   descriptor.Description,
+			Aliases: descriptor.Aliases,
+		},
+	}
+
+	// TODO(stmcginnis): There is a docs command we may want to add, need to
+	// see if that's necessary.
```

Comment:
The cobra docs command is really nice, especially for auto generating `man` pages. Might be something to consider in the future!!